distributed/shuffle/shuffle_extension.py

```diff
@@ -0,0 +1,329 @@
+            workers,
+        )
+
+        # Start the shuffle on all peers
+        # Note that this will call `shuffle_init` on our own worker as well
+        await asyncio.gather(
+            *(
+                self.worker.rpc(addr).shuffle_init(metadata=to_serialize(metadata))
+                for addr in metadata.workers
+            ),
+        )
+        # TODO handle errors from peers, and cancellation.
+        # If any peers can't start the shuffle, tell successful peers to cancel it.
+
+        return metadata  # NOTE: unused in tasks, just handy for tests
+
+    def add_partition(self, data: pd.DataFrame, shuffle_id: ShuffleId) -> None:
+        sync(self.worker.loop, self._add_partition, data, shuffle_id)
+
+    async def _add_partition(self, data: pd.DataFrame, shuffle_id: ShuffleId) -> None:
+        """
+        Task: Hand off an input partition to the ShuffleExtension.
+
+        This will block until the extension is ready to receive another input partition.
+
+        Using an unknown ``shuffle_id`` is an error.
+        """
+        await self._get_shuffle(shuffle_id).add_partition(data)
+
+    def barrier(self, shuffle_id: ShuffleId) -> None:
+        sync(self.worker.loop, self._barrier, shuffle_id)
+
+    async def _barrier(self, shuffle_id: ShuffleId) -> None:
+        """
+        Task: Note that the barrier task has been reached (`add_partition` called for all input partitions)
+
+        Using an unknown ``shuffle_id`` is an error. Calling this before all partitions have been
+        added is undefined.
+        """
+        # NOTE: in this basic shuffle implementation, doing things during the barrier
+        # is mostly unnecessary. We only need it to inform workers that don't receive
+        # any output partitions that they can clean up.
+        # (Otherwise, they'd have no way to know if they needed to keep the `Shuffle` around
+        # for more input partitions, which might come at some point. Workers that _do_ receive
+        # output partitions could infer this, since once `get_output_partition` gets called the
+        # first time, they can assume there are no more inputs.)
+        #
+        # Technically right now, we could call the `shuffle_inputs_done` RPC only on workers
+        # where `metadata.npartitions_for(worker) == 0`.
+        # However, when we have buffering, this barrier step will become important for
+        # all workers, since they'll use it to flush their buffers and send any leftover shards
+        # to their peers.
+
+        metadata = self._get_shuffle(shuffle_id).metadata
+
+        # Set worker restrictions for unpack tasks
+
+        # Could do this during `create_shuffle`, but we might as well overlap it with the time
+        # workers will be flushing buffers to each other.
+        name = "shuffle-unpack-" + metadata.id  # TODO single-source task name
+
+        # FIXME TODO XXX what about when culling means not all of the output tasks actually exist??!
```

Comment:
I don't know what to do about the additional transfer but the cleanup/output task pruning could be handled here by accessing 
```python
worker = get_worker()
key = worker.get_current_task()
ts = worker.tasks[key]
# assert ts == barrier
ts.dependents
```

Reply:
I assume knowing which dependents belongs to which ID boils down to the same problem we have on the scheduler plugin side, wouldn't it? The only way to properly assign the subset of dependents would be to parse they key for it's ID? Is there any other method to do this properly upfront during setup?

Reply:
I don't have a very thorough understanding of the HLG and Layers classes but just looking over the API/signatures I would assume we were able to wrap the entire shuffle graph into a single layer and handle the culling there. That would allow us to mutate transfer and barrier tasks solving both problems.

```python
class Layer:
    def cull(...):
        ...
```

https://github.com/dask/dask/blob/ecddc8df5869679cef5280c869674cbf0d9f60d9/dask/highlevelgraph.py#L112-L156


Similarly, if we rather wanted to use resources, annotations or other attributes, the layer materialization would be an ideal place where we could set all of this, couldnt' we? We don't need any further infrastructure for this, do we?

Reply:
> the cleanup/output task pruning could be handled here by accessing `worker.tasks[key].dependents`

Maybe I don't know worker state well enough, but I didn't think the scheduler sent workers the full list of dependents for a task. Don't workers only fill in dependents based on the dependencies of other tasks they've been told about? And since none of the dependents of the barrier can run until the barrier completes, the scheduler won't have sent those tasks to any workers, so I'd assume `dependents` is empty at this moment.

If this isn't true, and workers have a wider view of the graph than just the tasks they've been told to run, then that would change things a lot.

> The only way to properly assign the subset of dependents would be to parse they key for it's ID?

You need to know which _output partition number_ that task is creating. By definition for DataFrames, the output partition number is always the `i` part of the key `(name, i)`. So that's what we parse out of the key.

> Is there any other method to do this properly upfront during setup?

You could annotate each task with the output partition number, as mentioned.

You could maybe, very sketchily, assume something about the ordering/priorities of the keys instead of parsing them. Obviously just sorting by key names would work in most cases, because the keys all have the `i` component, but that's really no different from parsing. I _think_ maybe `enumerate(sorted(output_tasks, key=lambda ts: ts.priority))` would happen to give you the output partition number correctly simply because `dask.order` uses the ordering of key names to break ties—so again, you're still using the information in the key names.

Going fully mad, you could deserialize and inspect the task itself, traversing through all the layers of SubraphCallables until you found a call to `unpack`, and look at the `i` argument passed into it.

The bottom line is that in dask/dask, key names are not opaque. In fact, they're the place where this information (the partition number) is _defined_. So it's not surprising that most things come back to key names.

Reply:
> wrap the entire shuffle graph into a single layer and handle the culling there

This is a reasonable idea, and I think how HLGs were intended work. There are already special layers for task-based shuffles; it makes sense there'd be one for this as well. And yeah, this would solve a lot of our problems around culling, annotations, etc. I avoided doing this initially because the whole thing is easily expressible just with Blockwise, and the Layer API is so ill-specified and a pain to work with.

The problem with this approach is that we'd lose fusion, because it wouldn't be a Blockwise layer anymore, so `optimize_blockwise` wouldn't know how to fuse it. And then that would expose us to root task overproduction for both the inputs to the shuffle and the outputs from it. We'd want to write our own optimization step for this layer type to fuse it to Blockwise layers (big pain), or extend `optimize_blockwise` to know about this layer (odd), or somehow make it a subclass of Blockwise (possibly odd and tricky, since Blockwise is very much just one logical layer, and this would be two Blockwises plus a materialized layer in between).

-------

Also, I hope this illustrates just how much we have to bend ourselves in knots because the system can't schedule linear chains of tasks properly (https://github.com/dask/distributed/issues/5555).

If we had STA (or any other solution to this problem), I don't think we'd even be considering this scheduler plugin approach. We'd just write a simple Layer here with a cull method, not worry about fusion, not worry about unstable task names or parsing keys, not worry about propagating annotations, and do all the necessary logic within tasks, with a couple RPCs to the scheduler to get peers and set restrictions/resources.